oclc/classify.py

```diff
@@ -26,15 +29,204 @@
 )
 from core.util import MetadataSimilarity
 from core.util.xmlparser import XMLParser
+from coverage_utils import MetadataWranglerBibliographicCoverageProvider
 from viaf import NameParser as VIAFNameParser
 
-
 class OCLC(object):
     """Repository for OCLC-related constants."""
     EDITION_COUNT = "OCLC.editionCount"
     HOLDING_COUNT = "OCLC.holdings"
     FORMAT = "OCLC.format"
 
+class OCLCClassifyXMLParser(XMLParser):
+    """Turn the output of the OCLC Classify API into a Metadata object.
+
+    This object includes ContributorData for the book's contributors,
+    ClassificationData representing the most common classifications for
+    the book, and MeasurementData describing how many libraries have
+    collected the book.
+    """
+
+    # OCLC in-representation 'status codes'
+    SINGLE_WORK_SUMMARY_STATUS = 0
+    SINGLE_WORK_DETAIL_STATUS = 2
+    MULTI_WORK_STATUS = 4
+    NO_INPUT_STATUS = 100
+    INVALID_INPUT_STATUS = 101
+    NOT_FOUND_STATUS = 102
+    UNEXPECTED_ERROR_STATUS = 200
+
+    INTS = set([OCLC.HOLDING_COUNT, OCLC.EDITION_COUNT])
+
+    NAMESPACES = {'oclc' : 'http://classify.oclc.org'}
+    ROLES = re.compile("\[([^]]+)\]$")
+    LIFESPAN = re.compile("([0-9]+)-([0-9]*)[.;]?$")
+
+    CLASSIFIERS = {"fast": Subject.FAST, "lcc": Subject.LCC, "ddc": Subject.DDC}
+
+    @classmethod
+    def parse(cls, _db, xml):
+        tree = etree.fromstring(xml, parser=etree.XMLParser(recover=True))
+        contributors = cls.contributors(_db, tree)
+        measurements = cls.measurements(_db, tree)
+        subjects = cls.subjects(_db, tree)
+
+        return Metadata(
+            data_source=DataSource.OCLC,
+            contributors=contributors,
+            measurements=measurements,
+            subjects=subjects,
+        )
```

Comment:
Before adding tests there's one more change I'd like you to make.

I asked the OCLC Classify API about 73000 ISBNs and saved the XML files. We only got three different 'status codes':

2 (Single work detail): 78%
4 (Multi work detail): 10%
102 (not found): 11%

So those are the three the parser needs to handle.

Here's a bit of code from the old OCLCXMLParser for extracting the response code. I don't recommend copying any of the rest of the code from that class, except for the response code constants.

```
        response = cls._xpath1(tree, "oclc:response")
        representation_type = int(response.get('code'))
```

Your `parse` method should return a response code and a _list_ of Metadata objects. If the response code is 0 or 2, the list will contain one Metadata object -- that's basically the case you handle now. If the response code is 4, the list will contain several Metadata objects (you have some of this code now; I'll mention it later). Otherwise the list will be empty.